File: parking-lot/java/ParkingLot.java
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class ParkingLot {
    private final List<Level> levels;

    public ParkingLot(List<Level> levels) {
        this.levels = levels;
    }

    public ParkingSpot parkVehicle(Vehicle v) {
        for (Level level : levels) {
            Map<VehicleType, Integer> avail = level.getAvailability();
            if (avail.getOrDefault(v.getType(), 0) > 0) {
                ParkingSpot spot = level.parkVehicle(v);
                if (spot != null) {
                    return spot;
                }
            }
        }
        return null;
    }

    public Vehicle unparkVehicle(String spotId) {
        for (Level level : levels) {
            Vehicle v = level.unparkVehicle(spotId);
            if (v != null) {
                return v;
            }
        }
        return null;
    }

    public Map<VehicleType, Integer> getTotalAvailability() {
        Map<VehicleType, Integer> totalAvail = new ConcurrentHashMap<>();
        for (Level level : levels) {
            Map<VehicleType, Integer> avail = level.getAvailability();
            for (Map.Entry<VehicleType, Integer> entry : avail.entrySet()) {
                totalAvail.put(entry.getKey(), totalAvail.getOrDefault(entry.getKey(), 0) + entry.getValue());
            }
        }
        return totalAvail;
    }

    public void printAvailability() {
        System.out.println("--- Current Parking Availability ---");
        Map<VehicleType, Integer> avail = getTotalAvailability();
        for (Map.Entry<VehicleType, Integer> entry : avail.entrySet()) {
            if (entry.getValue() > 0 || entry.getKey() != VehicleType.UNKNOWN) {
                System.out.println(entry.getKey() + " Spots: " + entry.getValue());
            }
        }
        System.out.println("------------------------------------");
    }
}

File: parking-lot/cpp/ParkingLot.cpp
#include "ParkingLot.hpp"
#include <iostream>

ParkingLot::ParkingLot(const std::vector<std::shared_ptr<Level>>& levels) : Levels(levels) {}

ParkingSpot* ParkingLot::ParkVehicle(Vehicle* v) {
    for (const auto& level : Levels) {
        auto avail = level->GetAvailability();
        if (avail[v->Type] > 0) {
            ParkingSpot* spot = level->ParkVehicle(v);
            if (spot != nullptr) {
                return spot;
            }
        }
    }
    return nullptr;
}

Vehicle* ParkingLot::UnparkVehicle(const std::string& spotID) {
    for (const auto& level : Levels) {
        Vehicle* v = level->UnparkVehicle(spotID);
        if (v != nullptr) {
            return v;
        }
    }
    return nullptr;
}

std::map<VehicleType, int> ParkingLot::GetTotalAvailability() const {
    std::map<VehicleType, int> totalAvail;
    for (const auto& level : Levels) {
        auto avail = level->GetAvailability();
        for (const auto& pair : avail) {
            totalAvail[pair.first] += pair.second;
        }
    }
    return totalAvail;
}

void ParkingLot::PrintAvailability() const {
    std::cout << "--- Current Parking Availability ---\n";
    auto avail = GetTotalAvailability();
    for (const auto& pair : avail) {
        if (pair.second > 0 || pair.first != VehicleType::Unknown) {
            std::cout << VehicleTypeToString(pair.first) << " Spots: " << pair.second << "\n";
        }
    }
    std::cout << "------------------------------------\n";
}
